kernel/events.c

```diff
@@ -98,11 +98,19 @@ static int event_walk_op(struct k_thread *thread, void *data)
 
 	if (are_wait_conditions_met(thread->events, event_data->events,
 				    wait_condition)) {
+
+		/*
+		 * Events create a list of threads to wake up. We do
+		 * not want z_thread_timeout to wake these threads; they
+		 * will be woken up by k_event_post_internal once they
+		 * have been processed.
+		 */
+		thread->prevent_timeout = true;
```

Comment:
Dumb question: can't you just use z_abort_timeout() here?  Do you need to re-enable the pre-existing timeout for some reason?  Doesn't look like it at first glance.  If so that would simplify things enormously.

Reply:
What a great idea! I think it would work in most cases, but still leave one small hole to patch.

1. Core 0 starts walking the waitq in k_sched_post_internal (sched_spinlock is held).
2. Core 1 enters sys_clock_announce() and calls the associated timeout handler--z_thread_timeout().
3. Since Core 0 holds sched_spinlock, core 1 spins waiting for it.
4. Core 0 could abort almost all the relevant timeouts, save for the one whose timeout handler is already in progress on Core 1.
5. However, it should be possible to have an early bail out condition in z_thread_timeout() to cover that small hole.